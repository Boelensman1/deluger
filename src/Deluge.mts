import got, { Got } from 'got'
import { CookieJar } from 'tough-cookie'

import type {
  ConfigResult,
  StatusResult,
  Torrent,
  TorrentHash,
  Torrents,
  TorrentsWithProps,
  DelugeAddTorrentOptions,
} from './interfaces/index.mjs'
import torrentProps from './torrentProps.mjs'

interface DelugeError {
  code: number
  message: string
}

interface DelugeResult<T> {
  error: null | DelugeError
  id: number
  result: T
}

type AddTorrentOptions = Partial<{
  filePriorities: number[]
  addPaused: boolean
  compactAllocation: boolean
  downloadLocation: string
  maxConnections: number
  maxDownloadSpeed: number
  maxUploadSlots: number
  maxUploadSpeed: number
  prioritizeFirstLastPieces: boolean
  moveCompleted: boolean
  moveCompletedPath: string
  preAllocateStorage: boolean
  sequentialDownload: boolean
  seedMode: boolean
  superSeeding: boolean
}>

const getAddOptions = (
  options: AddTorrentOptions,
): DelugeAddTorrentOptions => ({
  download_location: options.downloadLocation,
  add_paused: options.addPaused,
  move_completed: options.moveCompleted,
  move_completed_path: options.moveCompletedPath,
  max_download_speed: options.maxDownloadSpeed,
  max_upload_speed: options.maxUploadSpeed,
  max_connections: options.maxConnections,
  max_upload_slots: options.maxUploadSlots,
  prioritize_first_last_pieces: options.prioritizeFirstLastPieces,
  file_priorities: options.filePriorities,
  sequential_download: options.sequentialDownload,
  super_seeding: options.superSeeding,
  compact_allocation: options.compactAllocation,
  pre_allocate_storage: options.preAllocateStorage,
  seed_mode: options.seedMode,
})

export class Deluge {
  private counter = -1
  private client: Got

  constructor(
    hostname: string,
    private password: string,
    port = 8112,
  ) {
    // create client
    this.client = got.extend({
      prefixUrl: `${hostname}:${port}`,
      retry: { limit: 2 },
      timeout: {
        request: 10000,
      },
      cookieJar: new CookieJar(), // add cookie support
    })
  }

  public async authenticate(): Promise<boolean> {
    const res = await this.fetch('auth.login', [this.password], false)
    return res === true
  }

  private async fetch<T>(
    method: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    params: any[] = [],
    autoLogin = true,
  ): Promise<T> {
    const body = {
      method,
      params,
      id: (this.counter += 1),
    }

    const result = await this.client
      .post('json', {
        json: body,
      })
      .json<DelugeResult<T>>()

    if (result.error !== null) {
      // check if error was due to not being authenticated and if so, login
      if (result.error.code === 1 && autoLogin) {
        return this.authenticate().then(() => this.fetch(method, params, false))
      }
      throw new Error(`Deluge api gave error: ${result.error.message}`)
    }
    return result.result
  }

  /**
   * Add a torrent to deluge
   *
   * @returns {string} The hash of the torrent
   */
  public async addTorrent(
    torrent: Buffer,
    options: AddTorrentOptions = {},
  ): Promise<TorrentHash> {
    const params = [
      'filename.torrent', // not sure what this is used for
      Buffer.from(torrent).toString('base64'),
      getAddOptions(options),
    ]

    const result = await this.fetch<TorrentHash>(
      'core.add_torrent_file',
      params,
    )

    if (result === null || result.length !== 40) {
      throw new Error('Deluge API did not return an id')
    }
    return result
  }

  public getConfig(): Promise<ConfigResult> {
    return this.fetch<ConfigResult>('core.get_config')
  }

  public getStatus<T extends (keyof Torrent)[]>(
    properties: T = torrentProps as T,
  ): Promise<StatusResult<T>> {
    const params = [properties, []]
    return this.fetch<StatusResult<T>>('web.update_ui', params)
  }

  public async listTorrents(torrentHashes?: TorrentHash[]): Promise<Torrents>
  public async listTorrents<T extends (keyof Torrent)[] = typeof torrentProps>(
    torrentHashes: TorrentHash[],
    properties: T,
  ): Promise<TorrentsWithProps<T>>
  public async listTorrents<T extends (keyof Torrent)[] = typeof torrentProps>(
    torrentHashes: TorrentHash[] = [],
    properties: T = torrentProps as T,
  ): Promise<Torrents | TorrentsWithProps<T>> {
    const params = [{}, properties]

    if (torrentHashes.length > 0) {
      params[0] = { id: torrentHashes }
    }

    return this.fetch('core.get_torrents_status', params)
  }

  public async removeTorrent(
    torrentHash: TorrentHash,
    removeData = false,
  ): Promise<boolean> {
    const params = [torrentHash, removeData]
    return this.fetch<boolean>('core.remove_torrent', params)
  }
}
