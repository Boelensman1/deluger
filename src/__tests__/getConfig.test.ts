import nock from 'nock'
import Deluge from '../index.js'

nock.load('./src/__tests__/responses/getConfig.json')

const deluge = new Deluge('http://host.com', 'passwd', 8083)

it('Gets the current status of the deluge client', () =>
  deluge.getConfig().then((config) => {
    expect(config).toBeDefined()
    expect(config).toMatchSnapshot()
  }))
