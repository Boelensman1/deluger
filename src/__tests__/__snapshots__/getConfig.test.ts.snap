// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gets the current status of the deluge client 1`] = `
Object {
  "add_paused": false,
  "allow_remote": true,
  "auto_managed": true,
  "autoadd_enable": false,
  "autoadd_location": "/Downloads",
  "cache_expiry": 60,
  "cache_size": 512,
  "compact_allocation": false,
  "copy_torrent_file": false,
  "daemon_port": 58846,
  "del_copy_torrent_file": false,
  "dht": false,
  "dont_count_slow_torrents": false,
  "download_location": "/mnt/data",
  "enabled_plugins": Array [
    "Label",
    "AutoAdd",
    "WebUi",
    "webapi",
  ],
  "enc_in_policy": 1,
  "enc_level": 2,
  "enc_out_policy": 1,
  "enc_prefer_rc4": true,
  "geoip_db_location": "/usr/local/share/GeoIP/GeoIP.dat",
  "ignore_limits_on_local_network": true,
  "info_sent": 0,
  "listen_interface": "",
  "listen_ports": Array [
    6870,
    6871,
  ],
  "lsd": true,
  "max_active_downloading": 5,
  "max_active_limit": 600,
  "max_active_seeding": 500,
  "max_connections_global": 350,
  "max_connections_per_second": 20,
  "max_connections_per_torrent": -1,
  "max_download_speed": 8000,
  "max_download_speed_per_torrent": -1,
  "max_half_open_connections": 50,
  "max_upload_slots_global": 4,
  "max_upload_slots_per_torrent": -1,
  "max_upload_speed": 10000,
  "max_upload_speed_per_torrent": -1,
  "move_completed": false,
  "move_completed_path": "/Downloads",
  "natpmp": true,
  "new_release_check": false,
  "outgoing_ports": Array [
    0,
    0,
  ],
  "peer_tos": "0x00",
  "plugins_location": "/home/deluge/.config/deluge/plugins",
  "prioritize_first_last_pieces": false,
  "proxies": Object {
    "dht": Object {
      "hostname": "",
      "password": "",
      "port": 8080,
      "type": 0,
      "username": "",
    },
    "peer": Object {
      "hostname": "",
      "password": "",
      "port": 8080,
      "type": 0,
      "username": "",
    },
    "tracker": Object {
      "hostname": "",
      "password": "",
      "port": 8080,
      "type": 0,
      "username": "",
    },
    "web_seed": Object {
      "hostname": "",
      "password": "",
      "port": 8080,
      "type": 0,
      "username": "",
    },
  },
  "queue_new_to_top": false,
  "random_outgoing_ports": true,
  "random_port": false,
  "rate_limit_ip_overhead": true,
  "remove_seed_at_ratio": true,
  "seed_time_limit": 180,
  "seed_time_ratio_limit": 7,
  "send_info": false,
  "share_ratio_limit": 2.5,
  "stop_seed_at_ratio": true,
  "stop_seed_ratio": 2.5,
  "torrentfiles_location": "/Downloads",
  "upnp": true,
  "utpex": false,
}
`;
